### PR TITLE
Submesh support

### PR DESCRIPTION
**Pull request by jpdean**
This PR adds the ability to create a submesh consisting of a subset of entities of an existing mesh. It is a step towards https://github.com/FEniCS/dolfinx/projects/9

Any feedback would be much appreciated!

### CODE REVIEW

File: cpp/dolfinx/mesh/Mesh.h [hunk 1]
```diff
@@ -101,6 +101,16 @@ class Mesh
...
 
+  /// Create a submesh consisting of a subset of entities in the
+  /// mesh (this)
+  /// @param[in] dim Entity dimension
+  /// @param[in] entities Vector of entities in the mesh
+  /// @return The submesh, a map from the (local) submesh vertices
+  /// to the (local) mesh vertices, and a map from the (local)
+  /// submesh geometry dofs to the (local) mesh geometry dofs
+  std::tuple<Mesh, std::vector<std::int32_t>, std::vector<std::int32_t>>
+  create_submesh(int dim, const xtl::span<const std::int32_t>& entities);
```
Comment (garth-wells) [inline cpp/dolfinx/mesh/Mesh.h:112]: Does this need to be a class method, or could it be a free function? Free functions are generally preferable. Methods are appropriate when access to private data is required.

File: cpp/dolfinx/mesh/Mesh.cpp [hunk 2]
```diff
@@ -171,7 +172,202 @@ Mesh mesh::create_mesh(MPI_Comm comm,
...
+
+  // Create submesh geometry index map
+  std::pair<common::IndexMap, std::vector<int32_t>> submesh_x_dof_index_map_pair
+      = geometry_dof_index_map->create_submap(submesh_owned_x_dofs);
+  auto submesh_x_dof_index_map = std::make_shared<common::IndexMap>(
+      std::move(submesh_x_dof_index_map_pair.first));
+
+  // Create a map from the (local) geometry dofs in the submesh to the (local)
+  // geometry dofs in the mesh (this).
+  std::vector<int32_t> submesh_to_mesh_x_dof_map;
```
Comment (garth-wells) [inline cpp/dolfinx/mesh/Mesh.cpp:296]:
```cpp
std::vector<int32_t> submesh_to_mesh_x_dof_map(submesh_owned_x_dofs.begin(), submesh_owned_x_dofs.end());
```

File: cpp/dolfinx/mesh/Mesh.cpp [hunk 3]
```diff
@@ -171,7 +172,202 @@ Mesh mesh::create_mesh(MPI_Comm comm,
...
+                                   submesh_owned_x_dofs.begin(),
+                                   submesh_owned_x_dofs.end());
+  for (auto x_dof_index : submesh_x_dof_index_map_pair.second)
+  {
+    int32_t x_dof = geometry_dof_index_map->size_local() + x_dof_index;
+    submesh_to_mesh_x_dof_map.push_back(x_dof);
+  }
+
+  // Create submesh geometry coordinates
+  xtl::span<const double> x = geometry().x();
```
Comment (garth-wells) [inline cpp/dolfinx/mesh/Mesh.cpp:309]: Would it be preferable to share the `x` data between meshes?

File: python/dolfinx/mesh.py [hunk 4]
```diff
@@ -54,6 +54,18 @@ def ufl_domain(self) -> ufl.Mesh:
         """Return the ufl domain corresponding to the mesh."""
         return self._ufl_domain
 
+    def create_submesh(self, dim, entities):
+        submesh, vertex_map, geom_map = self.create_submesh_cpp(dim, entities)
+        submesh_ufl_cell = ufl.Cell(submesh.topology.cell_name(),
+                                    geometric_dimension=submesh.geometry.dim)
+        # FIXME Don't hard code degree (and maybe Lagrange?)
+        submesh_domain = ufl.Mesh(
+            ufl.VectorElement("Lagrange",
```
Comment (garth-wells) [inline python/dolfinx/mesh.py:63]: Move calls onto one line for readability.

File: python/test/unit/mesh/test_mesh.py [hunk 5]
```diff
@@ -405,3 +407,111 @@ def test_unit_hex_mesh_assemble():
     vol = assemble_scalar(1 * dx(mesh))
     vol = mesh.comm.allreduce(vol, MPI.SUM)
     assert vol == pytest.approx(1, rel=1e-9)
+
+
+def boundary_0(x):
+    lr = np.logical_or(np.isclose(x[0], 0.0),
```
Comment (garth-wells) [inline python/test/unit/mesh/test_mesh.py:413]: Use longer lines in this function for readability.

**Review by garth-wells — CHANGES_REQUESTED** [review]
Looking good. Here's a first round of comments.

File: cpp/dolfinx/common/utils.cpp [hunk 6]
```diff
@@ -0,0 +1,96 @@
+// Copyright (C) 2021 Joseph P. Dean
+//
+// This file is part of DOLFINx (https://www.fenicsproject.org)
+//
+// SPDX-License-Identifier:    LGPL-3.0-or-later
+
+#include "utils.h"
+#include <dolfinx/common/MPI.h>
+#include <dolfinx/graph/AdjacencyList.h>
+#include <iostream>
```
Comment (jhale) [inline cpp/dolfinx/common/utils.cpp:10]: Include necessary? And following lines.

File: cpp/dolfinx/common/utils.cpp [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+#include <vector>
+
+std::vector<int32_t> dolfinx::common::compute_owned_indices(
+    const xtl::span<const std::int32_t>& indices,
+    const dolfinx::common::IndexMap& index_map)
+{
+  // Split indices into those owned by this process and those that
+  // are ghosts. `ghost_indices` contains the position of the ghost
+  // in index_map->ghosts()
+  std::vector<std::int32_t> owned;
```
Comment (jhale) [inline cpp/dolfinx/common/utils.cpp:22]: Can you reserve these vectors with approximate or exact size?

File: cpp/dolfinx/common/utils.cpp [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+  std::vector<std::int32_t> ghost_indices;
+  for (std::size_t i = 0; i < indices.size(); ++i)
```
Comment (jhale) [inline cpp/dolfinx/common/utils.cpp:24]: Possible to implement this using std::copy/std::copy_if?

File: cpp/dolfinx/common/utils.cpp [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+                 {
+                   auto it = std::find(global_indices.begin(),
+                                       global_indices.end(), global_index);
+                   assert(it != global_indices.end());
+                   return std::distance(global_indices.begin(), it);
+                 });
+
+  // Sort `owned` and remove non-unique entries (we could have received
+  // the same ghost from multiple other processes)
+  std::sort(owned.begin(), owned.end());
```
Comment (jhale) [inline cpp/dolfinx/common/utils.cpp:92]: We have some routines based on radix sorts for doing sorts of integer types with lower algorithmic complexity and higher memory complexity than std::sort.

https://github.com/FEniCS/dolfinx/blob/main/cpp/dolfinx/common/sort.h

File: python/test/unit/mesh/test_mesh.py [hunk 10]
```diff
@@ -405,3 +408,108 @@ def test_unit_hex_mesh_assemble():
...
+            mesh_entity = entities[submesh_entity]
+            mesh_entity_vertices = mesh_e_to_v.links(mesh_entity)
+            for i in range(len(submesh_entity_vertices)):
+                assert(vertex_map[submesh_entity_vertices[i]]
+                       == mesh_entity_vertices[i])
+    else:
+        assert(submesh.topology.index_map(entity_dim).size_local == 0)
+
+
+def submesh_geometry_test(mesh, submesh, geom_map, entity_dim, entities):
```
Comment (jhale) [inline python/test/unit/mesh/test_mesh.py:498]: Would be good to have some tests that use a submesh in onwards functionality, e.g. MeshTags, FunctionSpace etc.

File: cpp/dolfinx/common/IndexMap.cpp [hunk 11]
```diff
@@ -146,6 +147,84 @@ compute_owned_shared(MPI_Comm comm, const xtl::span<const std::int64_t>& ghosts,
...
+  const graph::AdjacencyList<std::int64_t> data_out(std::move(ghosts_to_send),
+                                                    std::move(send_disp));
+
+  // Communicate ghosts on this process in `indices` back to their owners
+  const graph::AdjacencyList<std::int64_t> data_in
+      = dolfinx::MPI::neighbor_all_to_all(reverse_comm, data_out);
+
+  // Get the local index from the global indices received from other
+  // processes and add to `owned`
+  const std::vector<std::int64_t>& global_indices = map.global_indices();
```
Comment (garth-wells) [inline cpp/dolfinx/common/IndexMap.cpp:210]: It would probably be better to create a local-to-global `std::unordered_map` here.

**Review by garth-wells — APPROVED** [review]

File: cpp/dolfinx/common/utils.cpp (hunk 8 above)
Comment (jpdean) [reply cpp/dolfinx/common/utils.cpp]: I don't think so, but I've used `std::for_each` instead

**Review by jhale — APPROVED** [review]
Thanks for addressing my comments! Happy when you're happy to merge @jpdean.

**Comment by jpdean** [comment]
Thanks for the feedback!

**Pull request merged**